lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -256,63 +256,59 @@ def add_level_two_and_three_data(xml, amount, data)
         xml.customerReference data[:customer_reference] if data.include?(:customer_reference)
         xml.cardAcceptorTaxId data[:card_acceptor_tax_id] if data.include?(:card_acceptor_tax_id)
-
         {
-          sales_tax: 'salesTax',
+          tax_amount: 'salesTax',
           discount_amount: 'discountAmount',
           shipping_amount: 'shippingAmount',
           duty_amount: 'dutyAmount'
         }.each do |key, tag|
           next unless data.include?(key)
 
           xml.tag! tag do
-            data_amount = data[key].symbolize_keys
-            add_amount(xml, data_amount[:amount].to_i, data_amount)
+            add_amount(xml, data[key].to_i, data)
           end
         end
 
-        xml.discountName data[:discount_name] if data.include?(:discount_name)
-        xml.discountCode data[:discount_code] if data.include?(:discount_code)
+        # xml.discountName data[:discount_name] if data.include?(:discount_name)
+        # xml.discountCode data[:discount_code] if data.include?(:discount_code)
 
-        add_date_element(xml, 'shippingDate', data[:shipping_date]) if data.include?(:shipping_date)
+        # add_date_element(xml, 'shippingDate', data[:shipping_date]) if data.include?(:shipping_date)
 
-        if data.include?(:shipping_courier)
-          xml.shippingCourier(
-            data[:shipping_courier][:priority],
-            data[:shipping_courier][:tracking_number],
-            data[:shipping_courier][:name]
-          )
-        end
+        # if data.include?(:shipping_courier)
+        #   xml.shippingCourier(
+        #     data[:shipping_courier][:priority],
+        #     data[:shipping_courier][:tracking_number],
+        #     data[:shipping_courier][:name]
+        #   )
+        # end
 
         add_optional_data_level_two_and_three(xml, data)
 
-        if data.include?(:item) && data[:item].kind_of?(Array)
-          data[:item].each { |item| add_items_into_level_three_data(xml, item.symbolize_keys) }
-        elsif data.include?(:item)
-          add_items_into_level_three_data(xml, data[:item].symbolize_keys)
-        end
+        data[:item].each { |item| add_items_into_level_three_data(xml, item.symbolize_keys, data) } if data.include?(:item)
       end
 
-      def add_items_into_level_three_data(xml, item)
+      def add_items_into_level_three_data(xml, item, data)
         xml.item do
           xml.description item[:description] if item[:description]
           xml.productCode item[:product_code] if item[:product_code]
           xml.commodityCode item[:commodity_code] if item[:commodity_code]
           xml.quantity item[:quantity] if item[:quantity]
-
-          {
-            unit_cost: 'unitCost',
-            item_total: 'itemTotal',
-            item_total_with_tax: 'itemTotalWithTax',
-            item_discount_amount: 'itemDiscountAmount',
-            tax_amount: 'taxAmount'
-          }.each do |key, tag|
-            next unless item.include?(key)
```

Comment:
The old "each on a hardcoded hash" strategy checked to see if the param existed in `item`, at this `next` clause. That check appears to have been removed? is that cool?

Not for nothing but--while I see why you flattened the each out (there are two XML nodes that require special code) it's a little verbose. What would you think of keeping the original implementation for unitCost, itemTotalWithTax, itemDiscountAmount, and taxAmount... with less functionality, like I like how you moved `to_i` to the function that needs it... and then having special ones for `itemTotal` and `unitOfMeasure`

Reply:
I had to change this because Worldpay is particular about the order in which the xml attributes are added and i kept having to interrupt the hash to get the order "right"